group_project_v2/stage_components.py

```diff
@@ -254,23 +282,35 @@ def submission_review_view(self, context):
         # element
         return fragment
 
+    def _validate_upload(self, request):
+        if not self.stage.available_now:
+            template = messages.STAGE_NOT_OPEN_TEMPLATE if not self.stage.is_open else messages.STAGE_CLOSED_TEMPLATE
+            # 422 = unprocessable entity
+            return 422, {'result': 'error', 'message': template.format(action=self.stage.STAGE_ACTION)}
+
+        if not self.stage.is_group_member and not self.stage.is_admin_grader:
+            # 403 - forbidden
+            return 403, {'result': 'error', 'message': messages.NON_GROUP_MEMBER_UPLOAD}
+
+        try:
+            self.validator(request.params[self.upload_id].file)
+        except ValidationError as validationError:
+            message = validationError.message % validationError.params
+            # 400 - BAD REQUEST
+            return 400, {'result': 'error', 'message': message}
+
+        return None, None
```

Comment:
This might be something to address in a next iteration (given your time constraints): All other returns are valid status code. I think it would be best if either:

- The function that called `_validate_upload` decided which return code to send the user.
- `_validate_upload` had a default success HTTP status that got returned, instead of `None`.